ee/modules/500-operator-trivy/images/report-updater/cache/cache.go

```diff
@@ -0,0 +1,289 @@
+	}
+
+	err = json.Unmarshal(b, &dockerConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal docker config.json: %w", err)
+	}
+
+	if containerRegistry, ok = dockerConfig.Auths[registry]; !ok {
+		return nil, fmt.Errorf("failed to find auth config for bdu registry %s", registry)
+	}
+
+	if len(containerRegistry.Auth) == 0 {
+		return nil, fmt.Errorf("bdu registry auth config is empty")
+	}
+
+	decoded, err := base64.StdEncoding.DecodeString(containerRegistry.Auth)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode bdu registry auth config: %w", err)
+	}
+
+	auth := string(decoded)
+
+	if len(strings.Split(auth, ":")) < 2 {
+		return nil, fmt.Errorf("bdu registry auth config seems to be malformed, should have the following format: 'user:password'")
+	}
+
+	d := &VulnerabilityCache{
+		logger: logger,
+		Dictionary: VulnerabilityDictionary{
+			Data: make(map[string][]string),
+		},
+		sourceConfig: RegistryConfig{
+			registry:   registry,
+			repository: strings.Split(image, ":")[0],
+			tag:        strings.Split(image, ":")[1],
+			user:       strings.Split(auth, ":")[0],
+			password:   strings.Split(auth, ":")[1],
+		},
+	}
+
+	err = d.initDictionary()
+	if err != nil {
+		return nil, err
+	}
+
+	return d, nil
+}
+
+// think about healthz check
+func (c *VulnerabilityCache) Check() error {
+	if len(c.Dictionary.Data) == 0 {
+		return fmt.Errorf("BDU dictionary empty")
+	}
+	return nil
+}
+
+func (c *VulnerabilityCache) getDataFromImageDescriptors() error {
+	//download
+	c.logger.Println("downloading BDU image")
+	//create oras in-memory storage
+	store := memory.New()
+	ctx := context.Background()
+
+	//set target repository
+	repo, err := remote.NewRepository(c.sourceConfig.repository)
+	if err != nil {
+		return err
+	}
+
+	//set repository auth
+	repo.Client = &auth.Client{
+		Client: retry.DefaultClient,
+		Cache:  auth.DefaultCache,
+		Credential: auth.StaticCredential(c.sourceConfig.registry, auth.Credential{
+			Username: c.sourceConfig.user,
+			Password: c.sourceConfig.password,
+		}),
+	}
+
+	//copy requested image from remote repository to oras in-memory storage and save its descriptor
+	descriptor, err := oras.Copy(ctx, repo, c.sourceConfig.tag, store, c.sourceConfig.tag, oras.DefaultCopyOptions)
+	if err != nil {
+		return fmt.Errorf("renewing BDU failed: couldn't copy BDU image to memory: %w", err)
+	}
+
+	//get successor descriptors of the descriptor
+	successors, err := content.Successors(ctx, store, descriptor)
+	if err != nil {
+		return fmt.Errorf("renewing BDU failed: couldn't get descriptors from BDU image: %w", err)
+	}
+
+	//iterate over descriptors to get the ones with relevant MediaType
+	for _, descriptor := range successors {
+		switch descriptor.MediaType {
+		case tarGzMediaType:
+			readCloser, err := store.Fetch(ctx, descriptor)
+			if err != nil {
+				fmt.Errorf("renewing BDU failed: couldn't fetch tar archive: %w", err)
+			}
+			defer readCloser.Close()
```

Comment:
`defer` in `for loop` is not safe, would be better to wrap it with a function. At least anonymous one